src/domain/RBF/rbf.cpp

```diff
@@ -0,0 +1,742 @@
+
+    VecRestoreArrayRead(f, &fvals) >> utilities::PetscUtilities::checkError;
+    VecRestoreArray(rhs, &vals) >> utilities::PetscUtilities::checkError;
+
+    MatSolve(A, rhs, weights) >> utilities::PetscUtilities::checkError;
+
+    VecDestroy(&rhs) >> utilities::PetscUtilities::checkError;
+
+    // Now do the actual interpolation
+
+    // Get the cell center
+    DMPlexComputeCellGeometryFVM(dm, c, NULL, x0, NULL) >> utilities::PetscUtilities::checkError;
+
+    PetscInt p1 = PetscMax(RBF::polyOrder + 1, 1), dim = subDomain->GetDimensions();
+    PetscInt px, py, pz;
+    PetscReal *xp;
+    PetscMalloc1(dim * p1, &xp) >> utilities::PetscUtilities::checkError;
+
+    for (PetscInt d = 0; d < dim; ++d) {
+        x0[d] = xEval[d] - x0[d];  // Shifted center
+
+        // precompute powers
+        xp[d * p1 + 0] = 1.0;
+        for (px = 1; px < p1; ++px) {
+            xp[d * p1 + px] = xp[d * p1 + (px - 1)] * x0[d];
+        }
+    }
+
+    PetscReal interpVal = 0.0;
+    VecGetArray(weights, &vals) >> utilities::PetscUtilities::checkError;
+    for (i = 0; i < nCells; ++i) {
+        interpVal += vals[i] * RBFVal(dim, x0, &x[i * dim]);
+    }
+
+    // Augmented polynomial contributions
+    switch (dim) {
+        case 1:
+            for (px = 0; px < p1; ++px) {
+                interpVal += vals[i++] * xp[0 * p1 + px];
+            }
+            break;
+        case 2:
+            for (py = 0; py < p1; ++py) {
+                for (px = 0; px < p1 - py; ++px) {
+                    interpVal += vals[i++] * xp[0 * p1 + px] * xp[1 * p1 + py];
+                }
+            }
+            break;
+        case 3:
+            for (pz = 0; pz < p1; ++pz) {
+                for (py = 0; py < p1 - pz; ++py) {
+                    for (px = 0; px < p1 - py - pz; ++px) {
+                        interpVal += vals[i++] * xp[0 * p1 + px] * xp[1 * p1 + py] * xp[2 * p1 + pz];
+                    }
+                }
+            }
+            break;
+        default:
+            throw std::runtime_error("ablate::domain::RBF::Interpolate encountered an unknown dimension.");
+    }
+
+    VecRestoreArray(weights, &vals) >> utilities::PetscUtilities::checkError;
+    VecDestroy(&weights) >> utilities::PetscUtilities::checkError;
+    PetscFree(xp) >> utilities::PetscUtilities::checkError;
+
+    return interpVal;
+}
+
+/************ End Interpolation Code **********************/
+
+/************ Begin Ablate support Code **********************/
+// Return the range of DMPlex objects at a given depth in a subDomain and region. This is pulled from ablate::solver::Solver::GetRange
+//    which already has the subDomain and region information.
+//    Note: This seems like it should be in ablate::domain and the solver will call this. Need to talk to Matt M. about it.
+void RBF::GetRange(std::shared_ptr<ablate::domain::SubDomain> subDomain, const std::shared_ptr<ablate::domain::Region> region, PetscInt depth, ablate::solver::Range &range) const {
```

Comment:
I agree.  Lets move this to subdomain instance and have GetRange in solver call it.